src/analysis/retail/priest/shadow/modules/tier/Tier31ShadowPriest4Set.tsx

```diff
@@ -0,0 +1,191 @@
+import SPELLS from 'common/SPELLS';
+import TALENTS from 'common/TALENTS/priest';
+import { TIERS } from 'game/TIERS';
+import Analyzer, { Options, SELECTED_PLAYER } from 'parser/core/Analyzer';
+import Events, {
+  ApplyBuffEvent,
+  ApplyBuffStackEvent,
+  RefreshBuffEvent,
+  DamageEvent,
+  RemoveBuffEvent,
+} from 'parser/core/Events';
+import Abilities from 'parser/core/modules/Abilities';
+import EventHistory from 'parser/shared/modules/EventHistory';
+import SpellUsable from 'parser/shared/modules/SpellUsable';
+import BoringSpellValueText from 'parser/ui/BoringSpellValueText';
+import Statistic from 'parser/ui/Statistic';
+import STATISTIC_CATEGORY from 'parser/ui/STATISTIC_CATEGORY';
+import { calculateEffectiveDamage } from 'parser/core/EventCalculateLib';
+import ItemDamageDone from 'parser/ui/ItemDamageDone';
+
+const MAX_DURATION = 60000; //max duration of the buff
+
+class Tier31ShadowPriest4Set extends Analyzer {
```

Comment:
All this logic is a PRIME candidate for using an `EventLinkNormalizer` - link T31 buffremove to Shadow Crash cast and SWP cast. You can then listen only for those casts and detect if they were buffed by checking the link.

You could then also you a link normalizer to link SWP cast to SWP direct damage, and SC cast to SC damage (with a large forward buffer to account for travel time, this should be fine because SC has a cooldown). Follow the link from cast to damage tally right there.

Reply:
I'll start working on using that

Reply:
I can mock up what I'm thinking if you'd like

Reply:
Here's what I have - the link specs in your normalizer would be something like this:

```ts
{
    linkRelation: CONSUMES_T31_BUFF,
    linkingEventId: [TALENTS.SHADOW_CRASH_TALENT.id, SPELLS.SHADOW_WORD_PAIN.id],
    linkingEventType: EventType.Cast,
    referencedEventId: SPELLS.SHADOW_PRIEST_TIER_31_4_SET_BUFF.id,
    referencedEventType: EventType.RemoveBuff,
    forwardBufferMs: BUFFER_MS,
    backwardBufferMs: BUFFER_MS,
    anyTarget: true,
  },
  {
    linkRelation: HITS_TARGET,
    linkingEventId: SPELLS.SHADOW_WORD_PAIN.id,
    linkingEventType: EventType.Cast,
    referencedEventId: SPELLS.SHADOW_WORD_PAIN.id,
    referencedEventType: EventType.Damage,
    forwardBufferMs: BUFFER_MS,
    backwardBufferMs: BUFFER_MS,
  },
  {
    linkRelation: HITS_TARGET,
    linkingEventId: TALENTS.SHADOW_CRASH_TALENT.id,
    linkingEventType: EventType.Cast,
    referencedEventId: TALENTS.SHADOW_CRASH_TALENT.id,
    referencedEventType: EventType.Damage,
    forwardBufferMs: SC_TRAVEL_BUFFER_MS,
    backwardBufferMs: BUFFER_MS,
    anyTarget: true,
  },
  ```
with constants
```ts
const BUFFER_MS = 100;
const SC_TRAVEL_BUFFER_MS = 3000; // long buffer to look for damage events to be safe - SC has much longer cooldown so this should be fine
```
and lookup helpers
```ts
export function consumedT31Buff(event: CastEvent): boolean {
  return HasRelatedEvent(event, CONSUMES_T31_BUFF);
}

export function getHits(event: CastEvent): DamageEvent[] {
  return GetRelatedEvents<DamageEvent>(event, HITS_TARGET);
}
```

Reply:
Sure, I haven't used a Normalizer before so a starting point would be great.

Reply:
Then in your actual analyzer you'd watch stack gains of the buff (because unfortunately you can't get prior stack count from the RemoveBuff) and also would watch casts of SWP and SC
```ts
  onBuffApplied(Event: ApplyBuffEvent) {
    this.stacks = 1;
  }

  onBuffStack(Event: ApplyBuffStackEvent) {
    this.stacks = Event.stack;
  }

onSwpCast(event: CastEvent) {
  if (consumedT31Buff(event)) {
    getHits(event).forEach(e -> {
        this.damageSWP += calculateEffectiveDamage(e, this.stacks * this.bonusSWP);
    }
  }
}

// similar to onSwpCast for onScCast
```

Reply:
Thank you!  I'll use all this as a starting point and get working on it

Reply:
It's possible the SWP hit linker also needs an `additionalCondition` to avoid accidentally tallying nearby ticks - not sure if this would practically happen tho

Reply:
I've got the normalizer working.  Thank you for your help on getting started with it, it ended up working great.
I did need to add an `additionalCondition` but it was only 100ms of ticks so it didn't make much of a difference.